DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp

```diff
@@ -5,140 +5,97 @@
 
-    h.single_task<Task<use_cache>>([=]() [[intel::kernel_args_restrict]] {
+    h.single_task<HistogramID<k_cache_depth>>(
+    [=]() [[intel::kernel_args_restrict]] {
 
       // On-chip memory for Histogram
-      uint32_t local_output[kNumOutputs];
-      uint32_t local_output_with_cache[kNumOutputs];
-
-      // Register-based cache of recently-accessed memory locations
-      uint32_t last_sum[kCacheDepth + 1];
-      uint32_t last_sum_index[kCacheDepth + 1];
-
-      // Initialize Histogram to zero
-      for (uint32_t b = 0; b < kNumOutputs; ++b) {
-        local_output[b] = 0;
-        local_output_with_cache[b] = 0;
-      }
-
+      // A k_cache_depth of 0 is equivalent to a standard array with no cache
+      fpga_tools::CachedLocalMemory<uint32_t, kNumOutputs, k_cache_depth> 
+        histogram(0);
       // Compute the Histogram
-      if (!use_cache) {  // Without cache
-        for (uint32_t n = 0; n < kInitNumInputs; ++n) {
-          // Compute the Histogram index to increment
-          uint32_t b = input[n] % kNumOutputs;
-          local_output[b]++;
-        }
-      } else {  // With cache
-
-        // Specify that the minimum dependence-distance of
-        // loop carried variables is kCacheDepth.
-        [[intel::ivdep(kCacheDepth)]] for (uint32_t n = 0;
-                                               n < kInitNumInputs; ++n) {
-          // Compute the Histogram index to increment
-          uint32_t b = input[n] % kNumOutputs;
-
-          // Get the value from the on-chip mem at this index.
-          uint32_t val = local_output_with_cache[b];
-
-          // However, if this location in on-chip mem was recently
-          // written to, take the value from the cache.
-          #pragma unroll
-          for (int i = 0; i < kCacheDepth + 1; i++) {
-            if (last_sum_index[i] == b) val = last_sum[i];
-          }
-
-          // Write the new value to both the cache and the on-chip mem.
-          last_sum[kCacheDepth] = local_output_with_cache[b] = val + 1;
-          last_sum_index[kCacheDepth] = b;
-
-          // Cache is just a shift register, so shift the shift reg. Pushing
-          // into the back of the shift reg is done above.
-          #pragma unroll
-          for (int i = 0; i < kCacheDepth; i++) {
-            last_sum[i] = last_sum[i + 1];
-            last_sum_index[i] = last_sum_index[i + 1];
-          }
-        }
+      for (uint32_t n = 0; n < kInitNumInputs; ++n) {
+        uint32_t hist_group = input[n] % kNumOutputs;
+        auto hist_count = histogram.read(hist_group);
+        hist_count++;
+        histogram.write(hist_group, hist_count);
       }
 
       // Write output to global memory
       for (uint32_t b = 0; b < kNumOutputs; ++b) {
-        if (!use_cache) {
-          output[b] = local_output[b];
-        } else {
-          output[b] = local_output_with_cache[b];
-        }
+        output[b] = histogram.read(b);
```

Comment:
You did a similar change below: you can rename `b` to `hist_group`

Reply:
Replaced all uses of b with hist_group